### ISSUE TITLE
Increase the Nonce in account contract

### ISSUE THREAD

**Issue by ltdai010** [OpenZeppelin/cairo-contracts#82]
When I try sending execute method, if the transaction is failed, the nonce is not update and I can't transfer again
- Problem
I tried creating a transaction and sent it. It's failed then I can't call the transaction anymore, unless I call another transaction to increase the nonce, then come back again
- Solution
So I think we can create an increase nonce method, to pass in the current nonce and update the nonce to the next nonce? So I can increase the nonce and call execute again.

**Comment by martriay** [comment]
If the transaction fails the nonce should not have changed. Adding an `increase_nonce` method wouldn't make sense, given that calling that method would have to be done through `execute` anyway, which already increases the nonce. Also if you transaction fails, this should fail too.

I think you are having another issue, can you share code and steps to reproduce the issue?

**Comment by ltdai010** [comment]
You can reproduce this by this way:
- sending a transaction with payload. But wrong signature.
- sending it again with the right signature, but it will not work. Because the payload is same lead to tx hash same(idk if StarkNet will change this), no way to recall the transaction then. Unless increase the nonce by calling another transaction.

**Comment by agolajko** [comment]
I bumped into the same problem today, and thought a bit about what is causing the issue. The issue seems to be that the [`execute`](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/Account.cairo#L140) command does not take a nonce from the user (there is a nonce argument but it's only used between the Account and the called contract, and has to take the value of [`_current_nonce`](https://github.com/OpenZeppelin/cairo-contracts/blob/main/contracts/Account.cairo#L164)).  So there is no nonce between the "wallet" and the account. Perhaps adding an extra nonce between the wallet and the account might fix this?

**Comment by hoangbuibunbu** [comment]
I think this problem must be fix by StarkNet itself. Adding timestamp or something to make the hash different each time we call.
Currently, I added increase nonce function. So when the transaction is failed, I can just increase the nonce to continue.

**Comment by ltdai010** [comment]
I think this problem should be fix by StarkNet itself. Adding timestamp or something to solve it.
Now, I just added increase nonce function in case anything go wrong.

**Comment by agolajko** [comment]
Yeah agree this seems to be on the plate of StarkNet to solve. Maybe doing what Ethereum mainnet is doing with the built in nonces would work. My current workaround (without changing the code) is passing different nonce values as inputs to the execute call and to the signature inside the call.

**Comment by bbrandtom** [comment]
Hi all,
I agree it's currently a problem. Solving the problem consist of two parts - ensuring transaction unicity on the protocol level, and splitting between transaction validation and transaction execution.
We are working atm on a new account abstraction model where those two topics are being handled and I think it will solve the issue you are describing here.
